---
layout: default
title: Research
---

We seek to learn how cells control the translation and stability of
mRNA transcripts and understand the role of this dynamic regulation in
maintaining homeostasis and adapting to changing environments. Cells
tightly control which genes they express as proteins, and the amount
of each protein they make, and post-transcriptional regulation can
modulate gene expression based on cellular needs and extracellular
conditions. Indeed, rapid reprogramming of protein synthesis in
dynamic environments often depends on modulating the translation and
stability of existing transcripts.

The first step in studying translational control is simply measuring
it accurately and completely.  I developed a ribosome profiling
technique that allows precise and quantitative analysis of genome-wide
in vivo translation. Ribosome profiling measures expression at the
level of translation, opening new windows on in vivo translation. It
also reports on the exact position of translating ribosomes. This
allows us to annotate the actual protein-coding sequences in
complicated genomes and transcriptomes. Ribosome profiling also
revealed extensive translation upstream of many genes.  This
unexpected translation points to the regulated use of alternate sites
of initiation, including substantial initiation at non-AUG codons.  We
believe that start codon selection is an underappreciated point of
control in gene expression.  The use of alternate sites of initiation
can produce functionally distinct protein isoforms as well as affect
the amount of protein synthesized.  We want to know the impact of this
phenomenon--when does the cell change start codon usage, and what are
the effects?

Regulation of mRNA translation and stability depends on mRNA-binding
proteins, which are as numerous as sequence-specific transcription
factors, suggesting a pervasive role for post- transcriptional
regulation in cell physiology. We are thus very interested in a
genetic and molecular understanding of these RNA-binding proteins, and
much of our work focuses on developing comprehensive and
high-throughput techniques for elucidating post-transcriptional
regulatory networks. 
